### ISSUE TITLE
Getting a higher single request throughput in rqlite 

### ISSUE THREAD

**Issue by rapoth** [rqlite/rqlite#380]
Thank you for your work on rqlite! I got it working on Windows without any issues. I have a few questions in terms obtaining higher single request throughput.

I started three local nodes:

```
.\rqlited.exe -on-disk ~/node.1
.\rqlited.exe -http-addr localhost:4003 -raft-addr localhost:4004 -join http://localhost:4001 -on-disk ~/node.2
.\rqlited.exe -http-addr localhost:4005 -raft-addr localhost:4006 -join http://localhost:4001 -on-disk ~/node.3
```

And then created a table:

```
CREATE TABLE test1(URI string);
```

And ran the benchmark:

```
.\rqbench.exe -x -b 1 -m 10 "INSERT INTO test1(URI) VALUES('hello');"
10 requests completed in 321.7902ms
...
90 requests completed in 320.9134ms
Total duration: 32.2257409s
Requests/sec: 3.10
Statements/sec: 3.10
```

I understand that I can get a much higher throughput using multiple requests in 1 batch --- e.g., I was seeing 48k/sec when I send in 50k statements in 1 batch, which is amazing. However, I 

I am curious as to what is causing this high single request latency. 

It looks like it takes ~300 ms/request on a local cluster of three nodes. If I count the number of communications (Client -> Leader (->Disk) -> Broadcast to followers TTL -> Apply to followers TTL), given that everything is happening on the same box, I am not understanding why it takes 300 ms for the entire run of the protocol. Are there any tweaks to improve this?

Thank you!

PS: Some ways I can think of doing this are (a) multiple clients issue requests at once, and (b) batch up requests. I will fallback to these in the worst case.

**Comment by otoolep** [comment]
Thanks for the report @rapoth -- the performance you see is similar to what I see. I do agree that 300ms looks rather long, but my testing shows it's the Raft use of the disks.

In my testing the bottlenecks are the disk, as the Raft system continually updates its log and fsyncs to disk after each change. You are also running in "on disk" mode which is causing more disk load. Unless you need that, you use an in-memory SQLite database. Your data is still safe.

Are you watching the CPU and disk load on your machine during your testing? You may see the disks being the bottleneck too. In that case I suggest you run a real 3-node cluster, because what you are doing during testing is causing 3x fsync traffic onto the single disk of the host machine -- . So for a more realistic performance testing of a 3-node cluster I suggest you do the following:

- do not enable "on-disk" mode.
- run your 3 nodes across 3 different machines.
- run the benchmarking tool from a 4th machine.
- have the best network you can (low latency).

Then see what kind of performance you get.

**Comment by rapoth** [comment]
@otoolep: Thank you for your prompt response! I will update this post with numbers from a cluster. In the meantime, some more details regarding my environment (along with a benchmark for in-memory sqlite):

  1. 48 cores, 128 GB RAM
  2. SSD 850 that can support upto 6 Gbps

I re-ran the benchmarks for sqlite in-memory mode (by restarting rqlite without any -on-disk parameter) and here's what I see:

```
PS E:\tmp\rqlite> .\rqbench.exe -x  -b 1 -m 10 "INSERT INTO test1(URI) VALUES('hello');"
10 requests completed in 310.5362ms
...
90 requests completed in 311.5332ms
Total duration: 31.1241688s
Requests/sec: 3.21
Statements/sec: 3.21
```
### CPU (1 statement/request)
![image](https://user-images.githubusercontent.com/8262876/32685448-87b48be2-c646-11e7-9d35-e96bee588fcc.png)

### SSD IO (1 statement/request)
![image](https://user-images.githubusercontent.com/8262876/32685429-df75374c-c645-11e7-9624-2d1cc1b0454e.png)

```
PS E:\tmp\rqlite> .\rqbench.exe -x -b 50000 -m 10 "INSERT INTO test1(URI) VALUES('hello');"
10 requests completed in 625.9143ms
...
90 requests completed in 663.6563ms
Total duration: 1m2.5478693s
Requests/sec: 1.60
Statements/sec: 79938.77
```

### CPU (50k statements/request)
![image](https://user-images.githubusercontent.com/8262876/32685466-bf043278-c646-11e7-94bb-bb908c65e0e8.png)

### SSD IO (50k statements/request)
![image](https://user-images.githubusercontent.com/8262876/32685418-a955bc90-c645-11e7-85d7-24e4147a3037.png)

In both cases, I am observing that drive IO is fairly low. The overall CPU did not go beyond 10%. 

I am currently setting up a cluster experiment and will follow-up. In the mean time, if you do need any other information, please let me know. I'm super interested in this project and am curious to see how we can obtain a higher throughput for 1 statement requests. Thank you for your time!

**Comment by otoolep** [comment]
OK, definitely interesting data -- thank you. I agree those numbers look kinda low, but I haven't studied the throughput issue in great detail. There may be room for improvement. Since I see my SSD 16-GB machine peg on IO, I left it at that. Perhaps there are other bottlenecks.

Are you familiar with profiling Go programs? I also suggest you re-run all testing passing `-cpu-profile` to each node, running the benchmark, and then shutting them down. There are details at https://blog.golang.org/profiling-go-programs that tell you what to do next. You could gather data and post it here.

Basically you'll do something like this:
```
.\rqlited.exe -cpu-profile cpu.prof ~/node.1
```
Terminate at the end of the test, and then execute:
```
go tool pprof rqlited cpu.prof
```
and run some of the commands in the blog post. I would be particularly interested in a profiling graph. I have not run these commands on Windows however, only Linux, but they should work.

**Comment by rapoth** [comment]
Thank you for the pointer. I am not familiar with Go at all but I will look at the blog and run the profiler. I will update this thread soon.

**Comment by rapoth** [comment]
I ran the profiler and collected the following. I am currently looking through them in more detail but I figured I'll attach them here in case you have any other ideas.

### node.1
[Profile Graph for node.1](http://svgshare.com/i/3r_.svg)
[Profile Graph (Malloc GC) for node.1](http://svgshare.com/i/3rR.svg)
```
PS E:\tmp\rqlite> go tool pprof rqlited.exe .\cpu.prof.1
Entering interactive mode (type "help" for commands)
(pprof) top10
1560ms of 1920ms total (81.25%)
Showing top 10 nodes out of 218 (cum >= 30ms)
      flat  flat%   sum%        cum   cum%
     770ms 40.10% 40.10%      770ms 40.10%  runtime.stdcall1
     520ms 27.08% 67.19%      520ms 27.08%  runtime.cgocall
      40ms  2.08% 69.27%       40ms  2.08%  runtime.lock
      40ms  2.08% 71.35%       40ms  2.08%  runtime.siftdownTimer
      40ms  2.08% 73.44%      260ms 13.54%  runtime.timerproc
      30ms  1.56% 75.00%       30ms  1.56%  runtime.acquirep
      30ms  1.56% 76.56%       40ms  2.08%  runtime.exitsyscall
      30ms  1.56% 78.12%       30ms  1.56%  runtime.runqgrab
      30ms  1.56% 79.69%      230ms 11.98%  runtime.startm
      30ms  1.56% 81.25%       30ms  1.56%  runtime.unlock

PS E:\tmp\rqlite>  go tool pprof rqlited.exe .\mem.prof.1
Entering interactive mode (type "help" for commands)
(pprof) top10
215.08kB of 269.41kB total (79.84%)
Dropped 210 nodes (cum <= 1.35kB)
Showing top 10 nodes out of 90 (cum >= 6.33kB)
      flat  flat%   sum%        cum   cum%
   67.04kB 24.89% 24.89%    67.04kB 24.89%  runtime.malg
   54.66kB 20.29% 45.17%    67.28kB 24.97%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).getConn
   16.14kB  5.99% 51.17%    88.53kB 32.86%  runtime.systemstack
   14.12kB  5.24% 56.41%    14.12kB  5.24%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.newNetPipeline
   12.86kB  4.77% 61.18%    12.86kB  4.77%  net.socket
   12.65kB  4.70% 65.88%    12.65kB  4.70%  github.com/boltdb/bolt.Open.func1
   12.14kB  4.51% 70.39%    12.14kB  4.51%  net.(*netFD).connect.func2
   11.41kB  4.23% 74.62%    11.41kB  4.23%  net/http.newBufioWriterSize
    7.72kB  2.87% 77.49%     7.72kB  2.87%  sync.(*Pool).pinSlow
    6.33kB  2.35% 79.84%     6.33kB  2.35%  github.com/boltdb/bolt.(*node).put

PS E:\tmp\rqlite>  go tool pprof --inuse_objects rqlited.exe .\mem.prof.1
Entering interactive mode (type "help" for commands)
(pprof) top10
1373 of 1440 total (95.35%)
Dropped 229 nodes (cum <= 7)
Showing top 10 nodes out of 71 (cum >= 43)
      flat  flat%   sum%        cum   cum%
       256 17.78% 17.78%        256 17.78%  github.com/hashicorp/go-msgpack/codec.(*ioDecReader).readn
       256 17.78% 35.56%        256 17.78%  runtime.gcBgMarkWorker
       236 16.39% 51.94%        410 28.47%  runtime.systemstack
       178 12.36% 64.31%        178 12.36%  runtime.malg
       129  8.96% 73.26%        129  8.96%  net.(*netFD).connect.func2
       128  8.89% 82.15%        128  8.89%  github.com/boltdb/bolt.(*Cursor).search
        61  4.24% 86.39%        118  8.19%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).getConn
        43  2.99% 89.38%         43  2.99%  github.com/hashicorp/go-msgpack/codec.NewDecoder
        43  2.99% 92.36%         43  2.99%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*netPipeline).decodeResponses
        43  2.99% 95.35%         43  2.99%  time.NewTimer
```

### node.2
[Profile Graph for node.2](http://svgshare.com/i/3pu.svg)
```
PS E:\tmp\rqlite>  go tool pprof rqlited.exe .\cpu.prof.2
Entering interactive mode (type "help" for commands)
(pprof) top10
630ms of 630ms total (  100%)
Showing top 10 nodes out of 98 (cum >= 10ms)
      flat  flat%   sum%        cum   cum%
     510ms 80.95% 80.95%      510ms 80.95%  runtime.cgocall
      60ms  9.52% 90.48%       60ms  9.52%  runtime.stdcall1
      10ms  1.59% 92.06%       10ms  1.59%  runtime.acquirep1
      10ms  1.59% 93.65%       10ms  1.59%  runtime.deferreturn
      10ms  1.59% 95.24%       10ms  1.59%  runtime.findrunnable
      10ms  1.59% 96.83%       70ms 11.11%  runtime.semawakeup
      10ms  1.59% 98.41%       80ms 12.70%  runtime.systemstack
      10ms  1.59%   100%       10ms  1.59%  runtime.timerproc
         0     0%   100%       10ms  1.59%  bufio.(*Reader).ReadByte
         0     0%   100%       10ms  1.59%  bufio.(*Reader).fill

PS E:\tmp\rqlite>  go tool pprof rqlited.exe .\mem.prof.2
Entering interactive mode (type "help" for commands)
(pprof) top10
1237.49kB of 1816.92kB total (68.11%)
Dropped 119 nodes (cum <= 9.08kB)
Showing top 10 nodes out of 90 (cum >= 62.86kB)
      flat  flat%   sum%        cum   cum%
  379.67kB 20.90% 20.90%   379.67kB 20.90%  github.com/boltdb/bolt.(*Tx).writeMeta
  237.47kB 13.07% 33.97%   237.47kB 13.07%  github.com/boltdb/bolt.(*node).put
  107.22kB  5.90% 39.87%   107.22kB  5.90%  github.com/boltdb/bolt.(*node).read
      96kB  5.28% 45.15%       96kB  5.28%  html.init
   84.82kB  4.67% 49.82%   326.04kB 17.94%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).handleCommand
   72.75kB  4.00% 53.82%   101.15kB  5.57%  github.com/hashicorp/go-msgpack/codec.(*encFnInfo).kStruct
   68.20kB  3.75% 57.58%   132.36kB  7.28%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeString
   64.34kB  3.54% 61.12%  1154.64kB 63.55%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*Raft).appendEntries
   64.16kB  3.53% 64.65%    64.16kB  3.53%  github.com/hashicorp/go-msgpack/codec.(*ioDecReader).readn
   62.86kB  3.46% 68.11%    62.86kB  3.46%  runtime.malg

PS E:\tmp\rqlite>  go tool pprof --inuse_objects rqlited.exe .\mem.prof.2
Entering interactive mode (type "help" for commands)
(pprof) top10
14572 of 21035 total (69.28%)
Dropped 127 nodes (cum <= 105)
Showing top 10 nodes out of 82 (cum >= 620)
      flat  flat%   sum%        cum   cum%
      3462 16.46% 16.46%       3462 16.46%  github.com/hashicorp/go-msgpack/codec.(*ioDecReader).readn
      3336 15.86% 32.32%       6798 32.32%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeString
      1586  7.54% 39.86%       8367 39.78%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*Raft).appendEntries
      1289  6.13% 45.99%      10380 49.35%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).handleCommand
      1024  4.87% 50.85%       1024  4.87%  github.com/boltdb/bolt.(*node).split
       903  4.29% 55.15%       1306  6.21%  github.com/hashicorp/go-msgpack/codec.(*encFnInfo).kStruct
       815  3.87% 59.02%       8511 40.46%  github.com/hashicorp/go-msgpack/codec.(*Decoder).decodeValue
       769  3.66% 62.68%        769  3.66%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).DecodePeer
       768  3.65% 66.33%        768  3.65%  github.com/mattn/go-sqlite3.(*SQLiteConn).prepare.func2
       620  2.95% 69.28%        620  2.95%  github.com/boltdb/bolt.(*Bucket).openBucket
```

### node.3
[Profile Graph for node.3](http://svgshare.com/i/3qE.svg)
```
PS E:\tmp\rqlite>  go tool pprof rqlited.exe .\cpu.prof.3
Entering interactive mode (type "help" for commands)
(pprof) top10
630ms of 630ms total (  100%)
Showing top 10 nodes out of 57 (cum >= 270ms)
      flat  flat%   sum%        cum   cum%
     460ms 73.02% 73.02%      460ms 73.02%  runtime.cgocall
     110ms 17.46% 90.48%      110ms 17.46%  runtime.stdcall1
      20ms  3.17% 93.65%       20ms  3.17%  runtime.unlock
      10ms  1.59% 95.24%      460ms 73.02%  internal/poll.(*FD).Fsync
      10ms  1.59% 96.83%       10ms  1.59%  math/rand.(*lockedSource).Int63
      10ms  1.59% 98.41%       10ms  1.59%  runtime.mcall
      10ms  1.59%   100%       10ms  1.59%  runtime.selectgo
         0     0%   100%      470ms 74.60%  github.com/boltdb/bolt.(*Tx).Commit
         0     0%   100%      200ms 31.75%  github.com/boltdb/bolt.(*Tx).write
         0     0%   100%      270ms 42.86%  github.com/boltdb/bolt.(*Tx).writeMeta

PS E:\tmp\rqlite>  go tool pprof rqlited.exe .\mem.prof.3
Entering interactive mode (type "help" for commands)
(pprof) top10
749.97kB of 1016.30kB total (73.79%)
Dropped 89 nodes (cum <= 5.08kB)
Showing top 10 nodes out of 73 (cum >= 24.10kB)
      flat  flat%   sum%        cum   cum%
  240.46kB 23.66% 23.66%   244.47kB 24.05%  github.com/boltdb/bolt.(*Tx).writeMeta
  139.94kB 13.77% 37.43%   139.94kB 13.77%  github.com/boltdb/bolt.(*node).put
   84.81kB  8.35% 45.78%   217.62kB 21.41%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).handleCommand
   63.49kB  6.25% 52.02%    63.49kB  6.25%  github.com/boltdb/bolt.(*node).read
   50.62kB  4.98% 57.00%    50.62kB  4.98%  github.com/boltdb/bolt.Open.func1
   44.45kB  4.37% 61.38%    48.47kB  4.77%  github.com/hashicorp/go-msgpack/codec.(*encFnInfo).kStruct
   40.24kB  3.96% 65.34%   712.24kB 70.08%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*Raft).appendEntries
   36.54kB  3.60% 68.93%   100.03kB  9.84%  github.com/boltdb/bolt.(*Bucket).node
   25.31kB  2.49% 71.42%   250.65kB 24.66%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).handleConn
   24.10kB  2.37% 73.79%    24.10kB  2.37%  github.com/boltdb/bolt.(*Tx).write

PS E:\tmp\rqlite>  go tool pprof --inuse_objects rqlited.exe .\mem.prof.3
Entering interactive mode (type "help" for commands)
(pprof) top10
6966 of 11118 total (62.66%)
Dropped 89 nodes (cum <= 55)
Showing top 10 nodes out of 73 (cum >= 342)
      flat  flat%   sum%        cum   cum%
      1298 11.67% 11.67%       5115 46.01%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).handleCommand
       898  8.08% 19.75%        898  8.08%  github.com/hashicorp/go-msgpack/codec.(*ioDecReader).readn
       898  8.08% 27.83%       1796 16.15%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeString
       877  7.89% 35.72%       5056 45.48%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*Raft).appendEntries
       727  6.54% 42.26%        727  6.54%  github.com/boltdb/bolt.(*Tx).write
       559  5.03% 47.28%        687  6.18%  github.com/hashicorp/go-msgpack/codec.(*encFnInfo).kStruct
       512  4.61% 51.89%        555  4.99%  github.com/mattn/go-sqlite3.(*SQLiteConn).begin
       512  4.61% 56.49%        512  4.61%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*NetworkTransport).DecodePeer
       343  3.09% 59.58%       1070  9.62%  github.com/rqlite/rqlite/vendor/github.com/hashicorp/raft.(*Raft).processHeartbeat
       342  3.08% 62.66%        342  3.08%  github.com/mattn/go-sqlite3.(*SQLiteStmt).exec
```

**Comment by otoolep** [comment]
Another interesting tool that might help diagnose performance issues: https://making.pusher.com/go-tool-trace/

**Comment by otoolep** [comment]
Related ideas in https://github.com/rqlite/rqlite/issues/469

**Comment by otoolep** [comment]
A complaint about rqlite performance (which is due to BoltDB): https://medium.com/@ivanjaros/event-store-with-boltdb-is-a-no-go-664b4781250d

The suggested alternatives for storage may be valid, and worth checking out.

**Comment by otoolep** [comment]
It might be time to try the Raft implementation from etcd: https://github.com/coreos/etcd/tree/master/raft

Perhaps its performance is better -- particularly its IO load.

**Comment by little-pan** [comment]
Using SQLite as the raft store instead of boltdb. Actually boltdb very poor performance.

**Comment by little-pan** [comment]
I'll provide raft-sqlite3 for the rqlite project😄

**Comment by little-pan** [comment]
I had printed time cost in (s *Service) handleExecute() of github.com/rqlite/http/service.go, the time cost about 1ms as time result(995.1µs or 1.0037ms etc) by timings parameter. But rqlite alway displays time cost 300ms above... Is overhead the http rqbench test?

**Comment by little-pan** [comment]
I think that I find where the issue lies in: **http.Client dns resolving problem**.
# Using IP address directly
```Go
>rqbench -a "127.0.0.1:4001" -b 1 -n 1 -m 1 "select *from accounts where id = 100000"
Total duration: 2.988ms
Requests/sec: 334.67
Statements/sec: 334.67
```
# Using default "localhost:4001"
```Go
>rqbench -b 1 -n 1 -m 1 "select *from accounts where id = 100000"
Total duration: 308.132ms
Requests/sec: 3.25
Statements/sec: 3.25
```
